clang/test/Driver/sycl-device-traits-macros-nvptx.cpp

```diff
@@ -0,0 +1,9 @@
+// REQUIRES: nvptx-registered-target
+
+// Check device traits macros are defined if sycl is enabled:
+// RUN:   %clangxx -fsycl -fsycl-targets=nvptx64-nvidia-cuda -### %s 2>&1 \
+// RUN:   | FileCheck -check-prefix=CHECK-SYCL-NVPTX-NVIDIA-CUDA %s
+// CHECK-SYCL-NVPTX-NVIDIA-CUDA-COUNT-2: "-D__SYCL_ANY_DEVICE_HAS_ANY_ASPECT__=1"
+// CHECK-SYCL-NVPTX-NVIDIA-CUDA: "-D__SYCL_ANY_DEVICE_HAS_{{.*}}__=1"
```

Comment:
That is my understanding too. Thanks for spotting this. The check line was carry-over from my amdgcn test and unfortunatelly the way it is written it passes the test despite it being ambiguous. I ought to remove it, speaking about:

```
// CHECK-SYCL-NVPTX-NVIDIA-CUDA: "-D__SYCL_ANY_DEVICE_HAS_{{.*}}__=1"
```

Cuda targets set `maySupportOtherAspects` to `true` due to stable backwards support of capabilities for Nvidia devices, so the same or newer devices will support the aspects of the former one. However, at the moment we cannot compile targeting multiple Cuda (or Hip) targets at the same, so it is a little difficult to fully test this stuff.

Reply:
Can we at least try different devices in several `RUN` lines? This is compile-only, so we don't need to have the actual device.

Reply:
I have updated the tests to not require a registered target to make these driver tests run regardless as they are checking driver logic anyways (I completely forgot there was `nogpulib`/`nocudalib` to avoid having to link device libraries etc).

I've added a couple more tests missing from yesterday that test the parsing more extensively for HIP and CUDA entries.

Don't think I can get -cc1 invocations for multiple nvidia or amd devices from one clang driver invocation because that is currently not implemented/supported in the driver. Basically `-fsycl-targets=nvidia_gpu_target1,nvidia_gpu_target2` will ignore the second nvidia device target and dispatch `JobAction`s only for the first target. That isn't something to worry about anyways and we can add a test for it in the parsing logic when that driver functionality gets extended. I know @jle-quel was looking into it, so please correct me if I am wrong.

So I think the tests are exhaustive atm, but I am open to any specific suggestions for further testing you may have @maarquitos14. I have addressed your first comment about the ambigious RUN check line regarding "`__SYCL_ANY_DEVICE_HAS_ANY_ASPECT__` is incompatible with any `__SYCL_ANY_DEVICE_HAS`". Thanks!

Reply:
I understand the limitation, but we can simply have multiple `RUN` lines compiling for a different target in every line, rather than one single line compiling for multiple targets. Does it make sense? I also left a comment pointing to the specific line I think we could repeat changing the target.